### PR TITLE
feat: add timeout to run host collector

### LINKED ISSUE #1377: Add a timeout to the run host collector
**Describe the rationale for the suggested feature.**

The [run host collector](https://troubleshoot.sh/docs/host-collect-analyze/run/) us used to run arbitrary command/programs which the collector has little control of with regards to how long the process takes. This can lead to the collector running for too long and in some cases, not even stopping

Collecting a support bundle using the spec below will never stop

```yaml
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  name: run
spec:
  hostCollectors:
    - run:
        name: "ping-google"
        command: "ping"
        args:
          - "google.com"
```

**Describe the feature**

Add a `timeout` field that can be used to limit how long the collector should run. The timeout should be
- A `string` [parsable into a time.Duration](https://pkg.go.dev/time#ParseDuration)
- Non negative. If its a negative value, default to passing no timeout

Here is an example spec

```yaml
    - run:
        name: "ping-google"
        timeout: 
        command: "ping"
        args:
          - "google.com"
```

**Describe alternatives you've considered**

Leave things as they are 

**Additional context**

- The run pod collector has a [timeout field](https://troubleshoot.sh/docs/collect/run-pod/#timeout-optional) which behaves as described above. This can be "borrowed"
- Eventually, [exec.CommandContext](https://pkg.go.dev/os/exec#CommandContext) would end up being called if there is a timeout

### PR DESCRIPTION
**Pull request by nvanthao**
## Description, Motivation and Context

#1377

## Checklist

- [X] New and existing tests pass locally with introduced changes.
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] The commit message(s) are informative and highlight any breaking changes
- [ ] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [X] No

### CODE REVIEW

File: pkg/collect/host_run.go [hunk 1]
```diff
@@ -49,7 +51,23 @@ func (c *CollectHostRun) Collect(progressChan chan<- interface{}) (map[string][]
...
+	var (
+		timeout            time.Duration
+		cmd                *exec.Cmd
+		errInvalidDuration error
+	)
+
+	ctx := context.Background()
+	cmdPath := c.attemptToConvertCmdToAbsPath()
+
+	timeout, errInvalidDuration = time.ParseDuration(runHostCollector.Timeout)
```
Comment (banjoh) [inline pkg/collect/host_run.go]: If the format is wrong, return an error like we do elsewhere

https://github.com/search?q=repo%3Areplicatedhq%2Ftroubleshoot%20%20time.ParseDuration&type=code

File: pkg/collect/host_run.go [hunk 2]
```diff
@@ -118,6 +136,9 @@ func (c *CollectHostRun) Collect(progressChan chan<- interface{}) (map[string][]
 		if werr, ok := err.(*exec.ExitError); ok {
 			runInfo.ExitCode = strings.TrimPrefix(werr.Error(), "exit status ")
 			runInfo.Error = stderr.String()
+		} else if err == context.DeadlineExceeded {
+			runInfo.ExitCode = "0"
```
Comment (banjoh) [inline pkg/collect/host_run.go]: Should't we use `werr.ExitCode()`? I'd prefer not "swallowing" the error code.

**Review by banjoh — COMMENTED** [review]
Changes look good. I added some minor comments

File: pkg/collect/host_run.go (hunk 1 above)
Comment (nvanthao) [reply pkg/collect/host_run.go]: thanks @banjoh! this is done!

File: pkg/collect/host_run.go (hunk 2 above)
Comment (nvanthao) [reply pkg/collect/host_run.go]: There is no Exit Code as as the error is about context becomes done prior to the command is run. Do you think we should update the artificial Exit Code to something else? E.g. `124`?

Comment (banjoh) [reply pkg/collect/host_run.go]: The value will be `-1` which the [std library uses to indicate](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/os/exec_posix.go;l=133) that the process did not complete. Any other value can be misinterpreted as a valid return code from the process.

**Comment by banjoh** [comment]
When you create a [PR to update the docs](https://github.com/replicatedhq/troubleshoot.sh/pulls), could you please add a one liner to indicate the `-1` exit code as an error possibly caused by timeouts?

File: pkg/collect/host_run.go (hunk 2 above)
Comment (nvanthao) [reply pkg/collect/host_run.go]: many thanks @banjoh this is done.

**Review by banjoh — APPROVED** [review]

**Pull request merged**